### PR TITLE
[java] Enable output pinning in OrtSession and OrtTrainingSession

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
This PR allows preallocated OrtValues to be supplied to the inference session run call, and the training session trainStep and evalStep calls. Combined with #16578 this allows users to have a zero copy path into and out of a run call, minimizing the number of copies and unnecessary allocations.

In the future we plan to use this as the basis for investigating zero copy paths using memory allocated on the GPU, and also to look exposing the existing ORT IO binding support in Java.

Note at the moment this exposes a possible bug in the type & shape validation of the inference session run call as it doesn't appear to check the tensor type of the supplied OrtValue before writing to it. The test code has been patched around it, but it can be triggered by changing "output-0" to "output-1" in line 855 of `InferenceTest.java`.

cc @yuslepukhin @baijumeswani 

### Motivation and Context
Improves performance by removing redundant copies.

### CODE REVIEW

**Comment by Craigacp** [comment]
Switching the requested output from "output-1" to "output-0" causes it to go through a different code path inside the CPU execution provider which does trigger an `ORT_RUNTIME_EXCEPTION` as expected, so the tests now pass.

**Comment by hariharans29** [comment]
Hi @yuslepukhin @baijumeswani - Can you please review this PR ?

File: java/src/main/native/ai_onnxruntime_OrtSession.c [hunk 1]
```diff
@@ -394,21 +405,26 @@ JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_run(JNIEnv* jniEnv
     goto cleanup_output_values;
   }
 
-  // Construct the output array of ONNXValues
-  jclass onnxValueClass = (*jniEnv)->FindClass(jniEnv, ORTJNI_OnnxValueClassName);
-  outputArray = (*jniEnv)->NewObjectArray(jniEnv, safecast_int64_to_jsize(numOutputs), onnxValueClass, NULL);
+  // Create the output boolean array denoting if ORT owns the memory for each output.
+  // Java boolean arrays are initialized to false.
+  outputArray = (*jniEnv)->NewBooleanArray(jniEnv, safecast_int64_to_jsize(numOutputs));
+  jboolean* boolArr = (*jniEnv)->GetBooleanArrayElements(jniEnv, outputArray, NULL);
```
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtSession.c:411]: >jboolean* boolArr [](http://example.com/codeflow?start=2&length=17)

The purpose of this array is not clear.

**Comment by yuslepukhin** [comment]
>   free(outputValues);

Is malloc() or calloc() being used to allocate this?

---
Refers to: java/src/main/native/ai_onnxruntime_OrtSession.c:431 in 8339bbf. [](commit_id = 8339bbf50b1dab29c5cdf8fcc4d0b00357e9ff1b, deletion_comment = False)

**Review by yuslepukhin — APPROVED** [review]
:shipit:

File: java/src/main/native/ai_onnxruntime_OrtSession.c (hunk 1 above)
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtSession.c:411]: It's a reference to the boolean array on the Java side which we use to track if this is a user supplied output or one ORT created in the run call. It makes the accounting a little easier back on the Java side as we write to the array of OnnxValues here on the C side so we needed either to copy it on the Java side and compare against null or allocate some other way of tracking it (which is what this is).

**Review by baijumeswani — DISMISSED** [review]

**Comment by Craigacp** [comment]
> > free(outputValues);
> 
> Is malloc() or calloc() being used to allocate this?
> 
> Refers to: java/src/main/native/ai_onnxruntime_OrtSession.c:431 in [8339bbf](https://github.com/microsoft/onnxruntime/commit/8339bbf50b1dab29c5cdf8fcc4d0b00357e9ff1b). [](commit_id = [8339bbf](https://github.com/microsoft/onnxruntime/commit/8339bbf50b1dab29c5cdf8fcc4d0b00357e9ff1b), deletion_comment = False)

It uses allocarray - https://github.com/microsoft/onnxruntime/pull/16835/files#diff-2ae00e781e96ba63ef1ff65ba6fc9ad1493b4a76f045c7ac9f8f18e3aa5ff8d9R363, which is either malloc or realloc - https://github.com/microsoft/onnxruntime/blob/main/java/src/main/native/OrtJniUtil.h#L101.

**Comment by hariharans29** [comment]
@Craigacp - it seems like some tests may be failing on Mac ?

**Comment by Craigacp** [comment]
> @Craigacp - it seems like some tests may be failing on Mac ?

I'm running it locally now, it might be because Dmitri added the extra validation and it changed the error code returned by the native code. If so I'll rebase on main and fix the exceptions we're checking in the tests.

**Comment by Craigacp** [comment]
Yep it was the native code validation changes. I moved all the checks over to invalid argument from runtime error and things pass for me locally on macOS ARM64. Should be good now.

**Comment by hariharans29** [comment]
azp run ONNX Runtime React Native CI Pipeline

**Review by hariharans29 — APPROVED** [review]

**Pull request merged**